### PR TITLE
Instaflux Demo App

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR introduces a new demo app called "Instaflux". It was meant to be a very simple Instagram like app which uses WordPress as it's back end. @tonyr59h and I have worked on this on a hackday and it was a very different app. Since we wanted something that resembles Instagram, I have decided to basically re-write this while also simplifying things.

This app will only work for a single site and for "image" posts. The only big missing feature I see with it is adding a title to the posts. I couldn't find an easy way to do it, so I decided to leave it for now. Note that if you upload a post from your browser with a title, it'll actually show up in the list. Also, if you have multiple images in a post, we'll just show the one with the largest width. I borrowed this idea from our reader.

Since this is basically a new app, the change set is quite large, I apologize for that in advance! Hope it'll be a useful demo for the project!

### CODE REVIEW

**Comment by oguzkocer** [comment]
Travis fails with a checkstyle error `You should not use _getSiteId_ in a XMLRPClient`. I have no idea why this is getting triggered as I am not using `getSiteId` on a `XMLRPClient` client as far as I can tell.

File: instaflux/src/main/res/values/styles.xml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        <item name="android:textColor">@android:color/white</item>
+    </style>
+
+    <style name="TextStyle" parent="TextAppearance.AppCompat">
+        <item name="android:textColor">@android:color/white</item>
+    </style>
+
+    <style name="EditTextStyle" parent="Widget.AppCompat.EditText">
+        <item name="backgroundTint">@android:color/white</item>
+        <item name="android:textColor">@android:color/white</item>
```
Comment (tonyr59h) [inline instaflux/src/main/res/values/styles.xml]: This should be black.

**Comment by tonyr59h** [comment]
That `find` command is just an extra step that Travis runs after `checkstyle`. It makes sure we use the correct site ID getter for self hosted sites. Travis failed the build due to the style errors above:

```[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/ImageUtils.java:18:27: '|' is not preceded with whitespace. [WhitespaceAround]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/ImageUtils.java:23:27: '|' is not preceded with whitespace. [WhitespaceAround]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/ImageUtils.java:23:28: '|' is not followed by whitespace. [WhitespaceAround]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/ImageUtils.java:28:27: '|' is not preceded with whitespace. [WhitespaceAround]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/ImageUtils.java:28:28: '|' is not followed by whitespace. [WhitespaceAround]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/PostActivity.java:54:23: Name 'MY_PERMISSIONS_READ_EXTERNAL_STORAGE' must match pattern '^m[a-zA-Z0-9]*$'. [MemberName]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/PostActivity.java:55:23: Name 'RESULT_PICK_MEDIA' must match pattern '^m[a-zA-Z0-9]*$'. [MemberName]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/PostAdapter.java:60:19: Name 'imageView' must match pattern '^m[a-zA-Z0-9]*$'. [MemberName]
[ant:checkstyle] [ERROR] /home/travis/build/wordpress-mobile/WordPress-FluxC-Android/instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/PostAdapter.java:61:18: Name 'titleTextView' must match pattern '^m[a-zA-Z0-9]*$'. [MemberName]```

**Comment by tonyr59h** [comment]
> The only big missing feature I see with it is adding a title to the posts.

If you set the post title to the media title in `onPostInstantiated` I think new posts will have a title. But for posts without a title you'd have to lookup the media by URL and set it for each post in `onPostChanged`.

**Comment by tonyr59h** [comment]
Nice work, it looks much nicer and does a good job of demoing the stores! I think once the checkstyle issues are resolved this one is good to merge.

P.S. 😆 @ the wp-stores icon making it this far.

**Comment by maxme** [comment]
> P.S. 😆 @ the wp-stores icon making it this far.

Hope you don't mind, I pushed a new icon ;)

ic_launcher-web

**Comment by maxme** [comment]
When I try to add a new picture, everything seems to work correctly but I end up with posts containing:

```
<img src="null" width="0" height="0" />
```

Example here: https://bia.is/wp2/?p=13

I checked my media library. picture is correctly uploaded.

File: instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/PostActivity.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            AppLog.w(AppLog.T.MEDIA, "OnMediaChanged error: " + event.error);
+        }
+    }
+
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onMediaUploaded(MediaStore.OnMediaUploaded event) {
+        if (event.progress >= 1.f && event.media != null) {
+            AppLog.i(AppLog.T.API, "Media uploaded: " + event.media.getTitle());
+            createMediaPost(event.media);
```
Comment (maxme) [inline instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/PostActivity.java]: `event.media` is an empty shell that only contains the remote mediaId. We should change this in the store, the `MediaStore.OnMediaUploaded ` should not contain a MediaModel, but just the id. Also `handleMediaUploaded` should write the updated media model in the DB when `event.completed` is true.

Comment (maxme) [inline instaflux/src/main/java/org/wordpress/android/fluxc/instaflux/PostActivity.java]: Used a work around in  	2054c32 to fix the issue, but we shouldn't pass "empty shell models" in the events.

**Comment by oguzkocer** [comment]
> That find command is just an extra step that Travis runs after checkstyle. It makes sure we use the correct site ID getter for self hosted sites. Travis failed the build due to the style errors above:

@tonyr59h we have already discussed these issues in detail on Slack :)

> If you set the post title to the media title in onPostInstantiated I think new posts will have a title. But for posts without a title you'd have to lookup the media by URL and set it for each post in onPostChanged.

@tonyr59h What I meant by missing title feature is the design, not the ability to pass it to server.

File: instaflux/src/main/res/values/styles.xml (hunk 1 above)
Comment (oguzkocer) [inline instaflux/src/main/res/values/styles.xml]: I'll look into dropping the styles completely.

**Comment by oguzkocer** [comment]
@maxme Thanks for the commits, I definitely don't mind 👍

> When I try to add a new picture, everything seems to work correctly but I end up with posts containing

I didn't notice this issue, thanks for the workaround. I'll look into this in a bit more detail.

**Comment by oguzkocer** [comment]
We decided to separate the self-hosted upload issues #257  since it's more on the store side than the demo app. With that, I think the PR is ready for another pass. Thanks!

**Comment by tonyr59h** [comment]
:shipit:

**Pull request merged**